carsus/io/output/base.py

```diff
@@ -48,89 +49,93 @@ def __init__(self,
+        self.gfall_reader = gfall_reader
+        self.zeta_data = zeta_data
+        self.chianti_reader = chianti_reader
+        self.cmfgen_reader = cmfgen_reader
+        self.levels_lines_param = levels_lines_param
+        self.collisions_param = collisions_param
 
-        # TODO: make this piece of code more readable 
-        gfall_ions = gfall_reader.levels.index.tolist()
-       
-        # Remove last element from tuple (MultiIndex has 3 elements)
-        gfall_ions = [x[:-1] for x in gfall_ions]
+        self.levels_all = self._get_all_levels_data()
+        self.lines_all = self._get_all_lines_data()
+        self.levels, self.lines = self.create_levels_lines(**levels_lines_param)
+        self.create_macro_atom()
+        self.create_macro_atom_references()
 
-        # Keep unique tuples, list and sort them
-        gfall_ions = sorted(list(set(gfall_ions)))
-        self.gfall_reader = gfall_reader
-        self.gfall_ions = gfall_ions
+        if (chianti_reader is not None) and (not chianti_reader.collisions.empty):
+            self.collisions = self.create_collisions(**collisions_param)
 
-        # TODO: priorities should not be managed by the `init` method.
-        self.chianti_reader = chianti_reader
-        if chianti_reader is not None:
-            chianti_lvls = chianti_reader.levels.reset_index()
-            chianti_lvls = chianti_lvls.set_index(
-                ['atomic_number', 'ion_charge', 'priority'])
+        logger.info('Finished.')
 
-            mask = chianti_lvls.index.get_level_values(
-                'priority') > self.gfall_reader.priority
+    @staticmethod
+    def solve_priorities(levels):
+        """ 
+        Returns a list of unique species per data source. 
+        
 
-            # TODO: make this piece of code more readable
-            chianti_lvls = chianti_lvls[mask]
-            chianti_ions = chianti_lvls.index.tolist()
-            chianti_ions = [x[:-1] for x in chianti_ions]
-            chianti_ions = sorted(list(set(chianti_ions)))
-            self.chianti_ions = chianti_ions
+        Notes
+        -----
 
-        else:
-            self.chianti_ions = []
+        The `ds_id` field is the data source identifier.
 
-        self.levels_all = self._get_all_levels_data()
-        self.lines_all = self._get_all_lines_data()
-        self._create_levels_lines(**self.levels_lines_param)
-        self._create_macro_atom()
-        self._create_macro_atom_references()
+        1 : NIST
+        2 : GFALL
+        3 : Knox Long's Zeta
+        4 : Chianti Database
+        5 : CMFGEN
 
-        if not self.chianti_reader.collisions.empty:
-            self.collisions = self._create_collisions()
+        """
+        levels = levels.set_index(['atomic_number', 'ion_number'])
+        levels = levels.sort_index()  # To supress warnings
 
-        self.zeta_data = zeta_data
+        lvl_list = []
+        for ion in levels.index.unique():
```

Comment:
what happens if they have the same priority? raise exception?

Reply:
No, in case of same priority `gfall` > `chianti` > `cmfgen` is assumed, I should put this in docs.

Reply:
Is this in the docs now? This is important information.

Reply:
No, I'll update the documentation in another PR